### PR TITLE
Initial development

### PR DESCRIPTION
**Pull request by lrezek**
Add the first version with basic features

### CODE REVIEW

**Review by RBowering — APPROVED** [review]

File: lib/leaf.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+	l.name = method.Call([]reflect.Value{})[0].String()
+}
+
+// initializeDependencies read in structure tags to find dependencies
+func (l *Leaf) initializeDependencies() {
+	l.unresolvedDependencies = map[string]reflect.Value{}
+	l.resolvedDependencies = map[string]reflect.Value{}
+
+	for i := 0; i < l.structureType.NumField(); i++ {
```
Comment (arianapeng) [inline lib/leaf.go:76]: Looks like you only loop through the first level of dependencies? What if A has a dependency on B, and then B has a dependency on C?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+func (s *SecondLeaf) GetLeafName() string {
+	return "SecondLeaf"
+}
+
+func (s *SecondLeaf) PostConstruct() {
+	fmt.Println("Second constructed, s.FirstLeaf is not nil here")
+}
+```
+
+You can now wire it them together:
```
Comment (arianapeng) [inline README.md:61]: wire ~it~ them together

File: README.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+```go
+package leaves
+
+first := &FirstLeaf{}
+second := &SecondLeaf{}
+
+tree := autumn.NewTree()
+tree.AddLeaf(FirstLeaf)
```
Comment (arianapeng) [inline README.md:69]: How does it work with interfaces?

File: lib/leaf.go (hunk 1 above)
Comment (lrezek) [reply lib/leaf.go:76]: If A has a dependency on B, the pointer to B will be injected in A. Once B is resolved, the pointer for C will be stuck into that B, which means that A.B.C will work once all deps have been resolved.

File: README.md (hunk 3 above)
Comment (lrezek) [reply README.md:69]: When adding the leaf, it has to be a structure pointer. However, when it's wired in, it should support interfaces as is 😄

**Pull request merged**